### ISSUE TITLE
Nextflow work folders for full-size datasets are HUGE! Can we slim it down?

### ISSUE THREAD

**Issue by FlicAnderson** [riboviz/riboviz#218]
After running some full-size datasets through Eddie, I've been able to get an idea of how much filesize expansion Nextflow is creating. 

Details of Cryptococcus H99 run [here](https://github.com/riboviz/example-datasets/issues/3#issuecomment-700142160) but the summary is:
 -  /input  27GB
 -  /output 9.4GB
 - /work  382GB
.... total size of the whole run folder: 791 GB

Given that scratch space is limited to 2TB quota, I can't run many full-size datasets on Eddie at once if they're all about this size when done, unless I throw a `rm -r /tmp` command on the end of my job submission script (would save 373GB in this case), and possibly delete the work folder to claw back around the same (382G for H99).  But I presume that deleting the work folder at the end of the job would mean I'd lose the `-resume` capacity, if part of the job didn't run correctly?  

Is there some way to easily 'slim down' the nextflow run a little?  As it stands, I definitely would start having trouble running things on my laptop if they're much bigger than this dataset...  

Certainly, it's a big plus for being able to run on Eddie and ignore this issue to some extent (if I run datasets one at a time), but it's possibly something we should bear in mind for non-cluster users, and if we want to think about running lots of samples from the same study, or even for meta-analyses of multiple datasets potentially?

**Comment by FlicAnderson** [comment]
Cryptococcus JEC21 Dataset disk usage: 

/input 18GB
/output 7.6GB
/work 267GB
.... total size of the whole run folder: 551 GB

Further info on this dataset [here](https://github.com/riboviz/example-datasets/issues/4#issuecomment-700161127)

**Comment by FlicAnderson** [comment]
Mike has clarified that the size issue is caused by copies of `/tmp` in `/work`.  That's why size of `/work = /tmp + /output`

**Comment by mikej888** [comment]
Explore use of `publishDir` directive (which currently copies files to `/tmp`) to publish symlinks or to do conditional copies (see for example ideas on https://gitter.im/nextflow-io/nextflow/archives/2016/11/09)

**Comment by ewallace** [comment]
Currently this is giving a footprint about 30x of input ( 791GB / 27GB = 29.2, 551GB / 18GB = 30.6).

Hopefully Mike's proposed fix will bring the footprint down to more like 15x of input.

**Comment by mikej888** [comment]
@FlicAnderson 

> But I presume that deleting the work folder at the end of the job would mean I'd lose the -resume capacity, if part of the job didn't run correctly?

Yes that's right. `-resume` uses intermediate results cached in `work/`.

**Comment by mikej888** [comment]
Commit 59f518f in nextflow-slim-218 branch:

* `publishDir` directive for processes producing index or temporary files sets its `mode` value based on `publish_index_tmp_type`.
* If the value of a newly-supported configuration parameter, `publish_index_tmp`, is true then `publish_index_tmp_type` is `copy` else it's `symlink.  

Note: `publishDir` does support an `enabled` parameter to contol whether to publish anything at all based on a boolean value. However, we need to publish either symlinks or file copies otherwise `countReads` fails since this invokes `riboviz.tools.count_reads` which, in turn, takes in the paths to the input, temporary and output directories. If it cannot find the files it needs to calculate the read counts then it fails with:

```
FileNotFoundError: [Errno 2] No such file or directory: '/home/ubuntu/riboviz/vignette/tmp'
```

This motivated the decision to use toggle the `publishDir` mode and use symbolic links by default.

Examples...

Specify `--publish_index_tmp` flag to publish index and temporary file copies:

```console
$ rm -rf vignette/index vignette/tmp vignette/output work
$ nextflow run prep_riboviz.nf -params-file vignette/vignette_config.yaml -ansi-log false --publish_index_tmp
$ du -sm vignette/index/ vignette/tmp/ vignette/output/ work/
9	vignette/index/
813	vignette/tmp/
3	vignette/output/
826	work/
$ pytest riboviz/test/regression/test_regression.py --expected=$HOME/vignette-20200708-03a4f13-nextflow  --skip-workflow --nextflow --check-index-tmp
============ 72 passed, 16 skipped, 1 warning in 143.85s (0:02:23) ============
$ rm -rf work
$ pytest riboviz/test/regression/test_regression.py --expected=$HOME/vignette-20200708-03a4f13-nextflow  --skip-workflow --nextflow --check-index-tmp
============ 72 passed, 16 skipped, 1 warning in 130.96s (0:02:10) ============
```

Regression tests for index and temporary files pass, as expected, as the files have been copied from `work/`.

Default, publish index and temporary file symlinks only:

```console
$ rm -rf vignette/index vignette/tmp vignette/output work
$ nextflow run prep_riboviz.nf -params-file vignette/vignette_config.yaml -ansi-log false
$ du -sm vignette/index vignette/tmp/ vignette/output/ work/
1	vignette/index
1	vignette/tmp/
3	vignette/output/
826	work/
```

Note the reduction in `tmp/` and `output/` space!

```console
$ pytest riboviz/test/regression/test_regression.py --expected=$HOME/vignette-20200708-03a4f13-nextflow  --skip-workflow --nextflow
================== 40 passed, 48 skipped, 1 warning in 1.65s ==================
$ pytest riboviz/test/regression/test_regression.py --expected=$HOME/vignette-20200708-03a4f13-nextflow  --skip-workflow --nextflow --check-index-tmp
============ 72 passed, 16 skipped, 1 warning in 143.85s (0:02:23) ============
```

Delete `work/`:

```console
$ rm -rf work
$ pytest riboviz/test/regression/test_regression.py --expected=$HOME/vignette-20200708-03a4f13-nextflow  --skip-workflow --nextflow --check-index-tmp
============ 32 failed, 40 passed, 16 skipped, 1 warning in 7.27s =============
$ pytest riboviz/test/regression/test_regression.py --expected=$HOME/vignette-20200708-03a4f13-nextflow  --skip-workflow --nextflow
```

Regression tests for index and temporary files now fail, as expected, as the symbolic links don't point to anything.

TODO:

* Add `publish_index_tmp: FALSE` to YAMLs.
* Update `riboviz.tools.upgrade_config.file`.
* Update docs.

**Comment by mikej888** [comment]
Completed:

* Add `publish_index_tmp: FALSE` to YAMLs.
* Update `riboviz.tools.upgrade_config.file`.
* Update docs.

**Comment by mikej888** [comment]
Created pull request #227.

**Comment by FlicAnderson** [comment]
After a super useful discussion with Mike Wallis & Angela Chitzanidi at the Eddie Users Drop-In, we've discussed Eddie space.

As a temporary measure until the end of November, the group space will have an increased quota - from 200GB to 5.2TB: 
ie. `/exports/csce/eddie/biology/groups/wallace_rna`. @ewallace has made this request, and agrees that we should discuss this space allocation towards the end of November so we can dial it back down again if necessary, as this seems to be quite easy to do.  

This will come in handy when any new honours students come online in January, as we can adjust the space as necessary for limited periods of time. 

So I can effectively now run datasets here as well as in my scratch space, and have some more room to play with if I'm running multiple datasets in parallel.

**Comment by ewallace** [comment]
We are closing this issue as it has been fixed by #227 , and increasing the `wallace_rna` quota on Eddie.
